functorch/functorch/experimental/cond.py

```diff
@@ -0,0 +1,133 @@
+    assert(isinstance(mode, TorchDispatchMode)), f"Unexpected mode type {mode.__class__}"
+    torch._C._set_torch_dispatch_mode(None)
+    try:
+        yield
+    finally:
+        torch._C._set_torch_dispatch_mode(mode)
+
+@contextmanager
+def enable_mode(mode):
+    curr_mode = torch._C._get_torch_dispatch_mode()
+    torch._C._set_torch_dispatch_mode(mode)
+    try:
+        yield
+    finally:
+        torch._C._set_torch_dispatch_mode(curr_mode)
+
+
+def trace_cond(proxy_mode, func_overload, pred, true_fn, false_fn, operands):
+    def _unwrap_proxy(e):
+        return get_proxy_slot(e, proxy_mode.tracer, e, lambda e: e.proxy)
+
+    assert isinstance(operands, list), "Cond operands must be a list of tensors"
+    assert all(isinstance(o, torch.Tensor) for o in operands), "Cond operands must be a list of tensors"
+
+    true_graph = get_isolated_graphmodule(true_fn, operands, {})
+    false_graph = get_isolated_graphmodule(false_fn, operands, {})
+
+    # There are probably better ways - I know that create_arg has some self incrementing name
+    # magic to it, but since we explicitly have to get the name for register_module,
+    # I was not sure how to do that. This kinda simulates it.
+    next_name = None
+    i = 0
+    while not next_name:
+        candidate = f"true_graph_{i}"
+        if hasattr(proxy_mode.tracer.root, candidate):
+            i += 1
+        else:
+            next_name = candidate
+
+    true_name = next_name
+    false_name = f"false_graph_{i}"
+    assert(not hasattr(proxy_mode.tracer.root, false_name))
+
+    proxy_mode.tracer.root.register_module(true_name, true_graph)
+    proxy_mode.tracer.root.register_module(false_name, false_graph)
+
+    # This is not amazing.
+    # However, if we have nested operators that have a call_function
+    # in their graph that is not a torch op (ex: see conditional below, nested cond)
+    # we cannot get metadata for it from just looking at out vars.
+    # The reason is that an operation on the output of such an op is not
+    # evalauted as a torch.Tensor.
+    # So we execute the real true and false fn here and compare metadata
+    true_result = true_graph(operands)
+    false_result = false_graph(operands)
+
+    def recursive_compare_same(a, b):
+        assert(type(a) == type(b))
+        if isinstance(a, torch.Tensor):
+            assert(a.dtype == b.dtype)
+            assert(a.size() == b.size())
+            assert(a.stride() == b.stride())
+            assert(a.device == b.device)
+        elif isinstance(a, (list, tuple)):
+            assert(len(a) == len(b))
+            for i in range(0, len(a)):
+                recursive_compare_same(a[i], b[i])
+
+    recursive_compare_same(true_result, false_result)
+
+    args = (pred, true_graph, false_graph, [operands])
+
+    proxy_args = pytree.tree_map(_unwrap_proxy, args)
+
+    return proxy_mode.tracer.create_proxy('call_function', func_overload, proxy_args, {},
```

Comment:
I'm a bit confused at how this works; the problem is probably just because I don't understand ProxyTensorMode end-to-end. Let's assume that we're tracing with concrete Tensors (so, `make_fx(f, mode='real')`) and we're using a simple `op = torch.ops.aten.sin`.
- During regular tracing, we get to `ProxyTensorMode.__torch_dispatch__(op, args, kwargs)`. [link](https://github.com/pytorch/pytorch/blob/247468baf0d9e105ced371a9eb42074147c1ee57/torch/fx/experimental/proxy_tensor.py#L332-L334)
- ProxyTensorMode executes `out = op(*args, **kwargs)` on real Tensors. ProxyTensorMode also executes `out_proxy = op(*proxy_args, *proxy_kwargs)`. Finally, ProxyTensorMode associates the `out` Tensor with the `out_proxy` Proxy. [link](https://github.com/pytorch/pytorch/blob/247468baf0d9e105ced371a9eb42074147c1ee57/torch/fx/experimental/proxy_tensor.py#L209-L218)
- `ProxyTensorMode.__torch_dispatch__` returns a Tensor. [link](https://github.com/pytorch/pytorch/blob/247468baf0d9e105ced371a9eb42074147c1ee57/torch/fx/experimental/proxy_tensor.py#L258)

Now, I am confused about why we're returning a Proxy here. I would have expected us to compute a real output tensor, then associate the output tensor with the proxy, and finally return the output tensor. Does that logic happen elsewhere?

Reply:
A more concrete example is the following. 
```py
import torch
from functorch import make_fx
from functorch.experimental.cond import cond

def f(x):
    y = x.sin()
    print(type(y))
    z = cond(y > 0, torch.sin, torch.cos, [y])
    print(type(z))
    w = z.sin()
    print(type(w))
    return w

x = torch.randn(3, 3)
make_fx(f)(x)
```
This prints
```
<class 'torch.Tensor'>
<class 'torch.fx.proxy.Proxy'>
<class 'torch.fx.proxy.Proxy'>
```
While tracing with real tensors, the intermediates should be real Tensors (as seen in the first print) but cond turns the intermediates into Proxy

Reply:
This is because we manually create a call_function - if we were to get a real tensor, it would mean we are getting it through executing a path, and then we would end up tracing path T or path F in the cond, all the way through, instead of storing a call out to the op in the graph. 

Does the graph code printout in the tests below help?